File: app/src/main/java/org/dhis2/utils/Result.java
package org.dhis2.utils;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import com.google.auto.value.AutoValue;

import java.util.ArrayList;
import java.util.List;

@AutoValue
public abstract class Result<T> {

    @NonNull
    public abstract List<T> items();

    @Nullable
    public abstract Exception error();

    @NonNull
    public static <E> Result<E> success(@NonNull List<E> items) {
        return new AutoValue_Result<>(items, null);
    }

    @NonNull
    public static Result failure(@NonNull Exception exception) {
        return new AutoValue_Result<>(new ArrayList<>(), exception);
    }
}

File: app/src/main/java/org/dhis2/utils/Result.kt
package org.dhis2.utils

import com.google.auto.value.AutoValue

import java.util.ArrayList

@AutoValue
abstract class Result<T> {

    abstract fun items(): List<T>

    abstract fun error(): Exception?

    companion object {

        fun <E> success(items: List<E>): Result<E> {
            return AutoValue_Result(items, null)
        }

        fun failure(exception: Exception): Result<*> {
            return AutoValue_Result(ArrayList<Any>(), exception)
        }
    }
}
